File: 11.sorting/selection_sort.go
package main

// func SelectionSortPseudoCode(input: []string) []string{
//   For an initial i position we have the current minimum value
//
//   minValue = input[i] // We need to know the min value in the subarray
//   minIndex = i // What we actually keep track of

//   for current_subarray_index in i+1 to n-1:
//     if input[current_subarray_index] < minValue:
//       minValue = input[current_subarray_index]
//       minIndex = current_subarray_index
//
// 			 Swap: input[i], input[minIndex]
// }

func SelectionSort(input []string) []string {
	for i := range input {
		smallestPosition := i
		for j := i + 1; j < len(input); j++ {
			if input[j] < input[smallestPosition] {
				smallestPosition = j
			}
		}
		input[i], input[smallestPosition] = input[smallestPosition], input[i]
	}
	return input
}

// func main() {
// 	input := []string{"b", "c", "d", "a", "a", "e", "f", "z", "a", "y", "w", "h", "g", "y"}

// 	sorted := SelectionSort(input)

// 	fmt.Printf("Selection Sort Result: %v\n", sorted)
// }

File: 11.sorting/selection_sort.py
class SelectionSort:
    def sort(self, input: list[str]) -> list[str]:
        for i in range(len(input)):
            smallest_position = i
            for j in range(i+1, len(input)):
                if input[j] < input[smallest_position]:
                    smallest_position = j
            input[i], input[smallest_position] = input[smallest_position], input[i]
        return input


s = SelectionSort()

input = ["b", "c", "d", "a", "e", "f", "z", "y", "w", "h", "g"]
print(f"Selection Sort Result: {s.sort(input)}")

input = [5, 8, 3, 9, 4, 1, 7]
print(f"Selection Sort Result: {s.sort(input)}")
